bibo/bibo.py

```diff
@@ -97,21 +97,31 @@ def _list_format_pattern(entries, format_pattern):
         click.echo(internals.format_entry(entry, format_pattern))
 
 
-def _list_citations(entries, database, bibstyle, verbose):
-    entries = [e for e in entries if e["type"] != "string"]
-    keys = [e["key"] for e in entries]
+def _list_citations(results, database, bibstyle, verbose):
+    results = list(results)
+    keys = [r.entry["key"] for r in results]
     exception = None
     try:
         citations = cite.cite(keys, database, bibstyle, verbose)
     except cite.BibtexException as e:
         exception = e
 
-    for entry in entries:
-        parts = [
-            internals.header(entry),
-            cite.fallback(entry) if exception else citations[entry["key"]],
-        ]
-        click.echo("\n".join(parts))
+    for result in results:
+        header = internals.header(result.entry)
+        if exception:
+            citation = cite.fallback(result.entry)
+        else:
+            citation = citations[result.entry["key"]]
+
+        text = "\n".join([header, citation])
+        text, extra_match_info = internals.highlight_match(text, result)
+
+        click.echo(text)
+
+        if extra_match_info:
+            click.secho("Search matched by", underline=True)
+        for key, val in extra_match_info.items():
+            click.echo("{}: {}".format(key, val))
```

Comment:
We can consider keeping only a few characters from each side of the highlight and truncate the rest. It might be better though to see how it performs at the moment and then decide.